### PR TITLE
Add support for non-analytic return for DefaultQubit.probability

### LINKED ISSUE #531: `default.qubit`  returns analytic probabilities when `analytic=False`
Currently, when `qml.probs()` is returned from a QNode, the `QubitDevice.probability()` abstract method is returned directly to the user.

For hardware-like devices that don't support analytic mode, this should behave as expected; the probability will be computed on the device by binning the generated computational basis samples.

However, for _analytic_ devices, such as `default.qubit`, care must be taken here. Currently, there is a bug in `DefaultQubit.probability()`, in that the analytic probability is _always_ returned, regardless what the value of `device.analytic` is:

```python
def probability(self, wires=None):
    if self._state is None:
        return None

    wires = wires or range(self.num_wires)
    prob = self.marginal_prob(np.abs(self._state) ** 2, wires)
    return prob
```

This should be modified as follows:
```python
def probability(self, wires=None):
    if self._state is None:
        return None

    wires = wires or range(self.num_wires)

    if self.analytic:
        prob = self.marginal_prob(np.abs(self._state) ** 2, wires)
        return prob

    # non-analytic mode, estimate the probability from the generated samples

    # consider only the requested wires
    wires = np.hstack(wires)
    samples = self._samples[:, np.array(wires)]

    # convert samples from a list of 0, 1 integers,
    # to base 10 representation
    unraveled_indices = [2] * len(wires)
    indices = np.ravel_multi_index(samples.T, unraveled_indices)

    # count the basis state occurrences, and construct
    # the probability vector
    basis_states, counts = np.unique(indices, return_counts=True)
    prob = np.zeros([len(wires)**2], dtype=np.float64)
    prob[basis_states] = counts/self.shots
    return prob
```

### PR DESCRIPTION
**Pull request by thisac**
**Context:**
Currently, when `qml.probs()` is returned from a QNode, the `QubitDevice.probability()` abstract method is returned directly to the user.

For hardware-like devices that don't support analytic mode, this should behave as expected; the probability will be computed on the device by binning the generated computational basis samples.

However, for analytic devices, such as default.qubit, care must be taken here. Currently, there is a bug in `DefaultQubit.probability()`, in that the analytic probability is always returned, regardless what the value of `device.analytic` is.

**Description of the Change:**
`DefaultQubit.probability()` is modified to return the non-analytic probability when called with `device.analytic = False`.

Furthermore, the analytic calculations of `DefaultQubit.probability()` is moved to a separate function called `DefaultQubit._analytic_probability()` which is called by `DefaultQubit.probability()` when `device.analytic = False` and from `QubitDevice.generate_samples()`, if available. If a QubitDevice has no `_analytic_probability()` function, then `QubitDevice.generate_samples()` will simply call the `probability()` function of the device directly.

**Benefits:**
`DefaultQubit.probability()` returns the correct probability when called with `device.analytic = False` or `device.analytic = True`.

**Possible Drawbacks:**
There is now an extra function called `_analytic_probabilities()`, that will only be called by `QubitDevice.generate_samples()` if that device has implemented it. Otherwise it will, as before, simply call `probabilities()` directly. This could cause some confusion (especially with plugins) whether to have this function or not.

**Related GitHub Issues:**
closes #531

### CODE REVIEW

File: pennylane/plugins/default_qubit.py [hunk 1]
```diff
@@ -218,10 +218,36 @@ def reset(self):
...
         if self._state is None:
             return None
 
         wires = wires or range(self.num_wires)
+
         prob = self.marginal_prob(np.abs(self._state) ** 2, wires)
         return prob
+
+    def probability(self, wires=None):
+        """Return the (marginal) probability of each computational basis."""
```
Comment (josh146) [inline pennylane/plugins/default_qubit.py]: As this method is defined in the parent abstract class, no need to define the docstring here 🙂 The docstring will automatically be inherited by the method.

This has several advantages:

* Reduces maintenance overhead --- docstring needs only be updated in one place

* Helps enforce the [Liskov substitution principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle), and avoids the temptation to modify the signature/docstring.

In limited cases, however, if the plugin _needs_ to add/change information in the docstring, it is okay to override to add this.

Comment (josh146) [reply pennylane/plugins/default_qubit.py]:
```suggestion
```

File: pennylane/_qubit_device.py [hunk 2]
```diff
@@ -280,7 +280,14 @@ def generate_samples(self):
...
         """
         number_of_states = 2 ** self.num_wires
-        rotated_prob = self.probability()
+
+        # Check if device has `_analytic_probability` implemented,
+        # otherwise call `probability`
+        try:
+            rotated_prob = self._analytic_probability()
+        except AttributeError:
+            rotated_prob = self.probability()
```
Comment (josh146) [inline pennylane/_qubit_device.py]: Nice!

File: tests/test_default_qubit.py [hunk 3]
```diff
@@ -1594,3 +1594,39 @@ def test_hermitian(self, theta, phi, varphi, monkeypatch, tol):
...
+
+class TestAnalyticBool:
+    """Test analytic and non-analytic for probability method"""
+
+    def mock_analytic_counter(self, wires=None):
+        self.analytic_counter += 1
+        return np.array([1, 0, 0, 0], dtype=float)
+
+    @pytest.mark.parametrize("analytic", [True, False])
+    def test_call_generate_samples(self, analytic, monkeypatch):
```
Comment (josh146) [inline tests/test_default_qubit.py]: Don't forget to add docstrings for the test functions!

File: tests/test_default_qubit.py [hunk 4]
```diff
@@ -1594,3 +1594,39 @@ def test_hermitian(self, theta, phi, varphi, monkeypatch, tol):
...
+        monkeypatch.setattr(dev, "_analytic_probability", self.mock_analytic_counter)
+
+        # generate samples from a qubit device with no '_analytic_probabiliy' function
+        dev._samples = mock_qubit_device.generate_samples()
+        dev.probability()
+
+        # should call `_analytic_probability` through `probability` once only if analytic=True
+        # and never through `generate_samples` since the `mock_qubit_device` has no
+        # `_analytic_probabiliy` function.
+        assert self.analytic_counter == analytic
```
Comment (josh146) [inline tests/test_default_qubit.py]: Nice unit tests!

File: pennylane/_qubit_device.py (hunk 2 above)
Comment (antalszava) [reply pennylane/_qubit_device.py]: Really nice find!

tl;dr: we might have to enforce that statevector simulators _do_ have an `_analytic_probability` defined if they also have an `analytic` attribute. My suggestion would be to throw an error if this is not the case. HW and hw simulators will have to anyways redefine this method.

_More details_
Analytic probabilities will be needed to compute non-analytic probabilities. Although this might be coded up in a single `probability` method in general, maybe it would be best if we distinguished the two approaches from the start for plugins and required that they were separated logically.

A way to distinguish state vector simulators from HW or hw simulators is checking if the `analytic` attribute is defined.

A cleaner solution in the long rung would be creating separate classes for different types of devices (something that came up previously as well).

Comment (antalszava) [reply pennylane/_qubit_device.py]: If we do decide to do something like this, then we'll have to update the `QubitDevice` docs accordingly :) (Even now, it might be worth doing so at least as a note, as this is a shared part for each device that subclasses `QubitDevice`)

File: pennylane/plugins/default_qubit.py [hunk 5]
```diff
@@ -218,10 +218,36 @@ def reset(self):
...
+        samples = self._samples[:, np.array(wires)]
+
+        # convert samples from a list of 0, 1 integers, to base 10 representation
+        unraveled_indices = [2] * len(wires)
+        indices = np.ravel_multi_index(samples.T, unraveled_indices)
+
+        # count the basis state occurrences, and construct the probability vector
+        basis_states, counts = np.unique(indices, return_counts=True)
+        prob = np.zeros([len(wires) ** 2], dtype=np.float64)
+        prob[basis_states] = counts / self.shots
```
Comment (antalszava) [inline pennylane/plugins/default_qubit.py]: Just for the clarity of the logic, it might be worth organizing this into a `_non_analytic_probability`. This would help us, in the long run, to review these parts of the codes more easily, and gives a clear view of how exactly the logic is branching.

The initial comment here, for example, would serve as a great one-liner in the docstring. :)

File: tests/test_default_qubit.py [hunk 6]
```diff
@@ -1594,3 +1594,39 @@ def test_hermitian(self, theta, phi, varphi, monkeypatch, tol):
...
+        self.analytic_counter = False
+
+        dev = qml.device("default.qubit", wires=2, analytic=analytic)
+        monkeypatch.setattr(dev, "_analytic_probability", self.mock_analytic_counter)
+
+        # generate samples through `generate_samples` (using '_analytic_probability')
+        dev.generate_samples()
+
+        # should call `_analytic_probability` once through `generate_samples`
+        assert self.analytic_counter == 1
```
Comment (antalszava) [inline tests/test_default_qubit.py]: This is a really nice test case! :blush:

File: tests/test_default_qubit.py [hunk 7]
```diff
@@ -1594,3 +1594,39 @@ def test_hermitian(self, theta, phi, varphi, monkeypatch, tol):
             )
         ) / 16
         assert np.allclose(var, expected, atol=tol, rtol=0)
+
+class TestAnalyticBool:
```
Comment (antalszava) [inline tests/test_default_qubit.py]: As the `QubitDevice.generate_samples` inherited method is being called, basically what is being tested here is that the `probability` method of defaultqubit integrates well with it. So imo these can be regarded as integration tests. (Although also in most cases integration tests in PL would include creating `QNode`s)
```suggestion
class TestProbabilityIntegration:
```

File: tests/test_default_qubit.py [hunk 8]
```diff
@@ -1594,3 +1594,39 @@ def test_hermitian(self, theta, phi, varphi, monkeypatch, tol):
...
+    """Test analytic and non-analytic for probability method"""
+
+    def mock_analytic_counter(self, wires=None):
+        self.analytic_counter += 1
+        return np.array([1, 0, 0, 0], dtype=float)
```
Comment (antalszava) [inline tests/test_default_qubit.py]: Cool! :)

**Comment by josh146** [comment]
@thisac, what is the status of this PR, is it ready for another round of review?

**Comment by thisac** [comment]
@josh146, I was planning to finish this later today. Sorry for the delay. The hackaton (and other things) came in the way last week.

File: pennylane/plugins/default_qubit.py (hunk 5 above)
Comment (thisac) [reply pennylane/plugins/default_qubit.py]: Is this something that we would prefer as standard then, i.e. having three functions (`probability`, `_analytic_probability` and `non_analytic_probability`)?

File: pennylane/_qubit_device.py (hunk 2 above)
Comment (thisac) [reply pennylane/_qubit_device.py]: I think that @josh146 mentioned that i might be quite a bit more work if making the `_analytic_probability` (and the potential `_non_analytic_probability` as you mentioned below) a requirement, especially with regards to the plugins. But maybe this is ok if only putting the requirement on simulators?

I'm guessing it's a better fit for a new PR (after some discussions, or even following an ADR).

File: pennylane/plugins/default_qubit.py (hunk 5 above)
Comment (antalszava) [reply pennylane/plugins/default_qubit.py:251]: @josh146 what would be your take here?

From my side, I would be in favour of that. Just because this separation in logic would arise in any state vector simulator that is created for any plugin, given that the device supports both `analytic=True` and `analytic=False`.

File: pennylane/_qubit_device.py (hunk 2 above)
Comment (antalszava) [reply pennylane/_qubit_device.py:289]: Having said that, this requirement would rather draw the attention of plugin writers to the fact, that this is how it works. Since this subtlety was missed in `default.qubit`, I think it would be nice if we somehow highlighted this case. Having a note in the docs could also work (if we'd rather not have a clear separation for convenience).

**Comment by josh146** [comment]
@antalszava, @thisac: just moving the conversation top level, since GitHub has a habit of sometimes making the line comments hard to find.

There are potentially two logic flows here, which is why this is a bit confusing:

1. The user requests an expectation value or variance of an observable.

   - `analytic=True`: `QubitDevice` calls the devices probability method.

     * Only supported on statevector devices.
     * This should return the **analytic probability**, and the expectation value is computed via `lambda @ probs`, where `lambda` is the observable eigenvalues.

   - `analytic=False`: `QubitDevice` calls `generate_samples` to extract finite number of samples from the device. The mean and the variance of the samples is then calculated.
     * Supported on both statevector and hardware-like devices.
     * For hardware-like device, samples returned directly. **No probability method is called**.
     * If a statevector device, `QubitDevice` generates samples via the **analytic probability**.

2. The user requests the probability on particular wires be returned.

   - `analytic=True`: return the *analytic probability* (only supported on statevector devices)
   
   - `analytic=False`: return an estimate of the probability, either calculated from samples, or returned from the device itself.

So looking over this, perhaps the solution is:

* Analytic devices **must** provide an `analytic_probability()` method. `QubitDevice.generate_samples()` calls `analytic_probability()` internally, as in this PR.

* Hardware devices **must** overwrite `generate_samples()`

* `QubitDevice._estimate_probability(wires=None)` is a private method containing general logic (i.e., the logic from this PR) for estimating probabilities from `self._samples`. This allows it to be easily applied to other plugins, such as [the `forest.wavefunction` device](https://www.vecteezy.com/vector-art/247806-parties-and-gatherings-illustration).

* `QubitDevice` has a new **built in** probability method:

  ```python
  def probability(wires=None):
      if self.analytic:
          return self.analytic_probability(wires=wires)
      
      return self._estimate_probability(wires=wires)
   ```

* Hardware devices may optionally overwrite `probability()` if they provide a more-efficient built-in method (e.g., qiskit devices return the probability without needing first generate samples).

---

### Summary

With these changes, from the perspective of plugin developers:

* Statevector devices must provide `analytic_probability()`

* Hardware-like devices must provide `generate_samples()` and may optionally provide `probability()`.

I think this is relatively clear for plugin developers, while also minimizing breaking changes (as only statevector devices will need to be updated), and moving common code into `QubitDevice`.

**Comment by thisac** [comment]
I've added an `_estimate_probability` method to `DefaultQubit` which, together with the `analytic_probability` and `probability` methods, can easily be transferred to `QubitDevice` in a future PR together with the changes suggested by @josh146 above.

I also added some `NOTE`s mentioning what should be changed/removed when implementing the above suggestions (e.g. the try-except statement in `QubitDevice.generate_samples`).

File: pennylane/_qubit_device.py [hunk 9]
```diff
@@ -280,7 +280,15 @@ def generate_samples(self):
              array[complex]: array of samples in the shape ``(dev.shots, dev.num_wires)``
         """
         number_of_states = 2 ** self.num_wires
-        rotated_prob = self.probability()
+
+        # Check if device has `analytic_probability` method, otherwise call `probability`
+        # NOTE: try-except statement should be removed when enforcing an `analytic_probabily`
```
Comment (antalszava) [inline pennylane/_qubit_device.py:285]: This looks good and would remind us to carry out this change! Suggesting to change it to `TODO` such that it could be picked up by `pylint` for example.
```suggestion
        # TODO: try-except statement should be removed when enforcing an `analytic_probabily`
```

File: pennylane/plugins/default_qubit.py [hunk 10]
```diff
@@ -240,10 +240,38 @@ def reset(self):
...
         if self._state is None:
             return None
 
         wires = wires or range(self.num_wires)
+
         prob = self.marginal_prob(np.abs(self._state) ** 2, wires)
         return prob
+
+    def _estimate_probability(self, wires=None):
+        # non-analytic mode, estimate the probability from the generated samples
```
Comment (antalszava) [inline pennylane/plugins/default_qubit.py:254]:
```suggestion
        """Return the estimated probability of each computational basis state using the generated samples."""
```

File: pennylane/plugins/default_qubit.py [hunk 11]
```diff
@@ -240,10 +240,38 @@ def reset(self):
         self._state[0] = 1
         self._pre_rotated_state = self._state
 
-    def probability(self, wires=None):
+    def analytic_probability(self, wires=None):
+        """Return the (marginal) analytic probability of each computational basis."""
```
Comment (antalszava) [inline pennylane/plugins/default_qubit.py:244]:
```suggestion
        """Return the (marginal) analytic probability of each computational basis state."""
```

File: pennylane/plugins/default_qubit.py [hunk 12]
```diff
@@ -240,10 +240,38 @@ def reset(self):
...
+        # count the basis state occurrences, and construct the probability vector
+        basis_states, counts = np.unique(indices, return_counts=True)
+        prob = np.zeros([len(wires) ** 2], dtype=np.float64)
+        prob[basis_states] = counts / self.shots
+        return prob
+
+    def probability(self, wires=None):
+        wires = wires or range(self.num_wires)
+
+        if hasattr(self, "analytic") and self.analytic:
```
Comment (antalszava) [inline pennylane/plugins/default_qubit.py:274]: `default.qubit` being a state vector simulator has the `analytic` attribute defined. As per our discussion, I'd be in for having `hasattr(self, "analytic")` as part of the condition once this is moved to `QubitDevice` (or leaving the try-except block seems a comparably good solution as well to me).
```suggestion
        if self.analytic:
```

File: pennylane/plugins/default_qubit.py [hunk 13]
```diff
@@ -240,10 +240,38 @@ def reset(self):
...
+        """Return the (marginal) analytic probability of each computational basis."""
         if self._state is None:
             return None
 
         wires = wires or range(self.num_wires)
+
         prob = self.marginal_prob(np.abs(self._state) ** 2, wires)
         return prob
+
+    def _estimate_probability(self, wires=None):
```
Comment (josh146) [inline pennylane/plugins/default_qubit.py:253]: Now that I think about it more, I'm not sure why I wrote `_estimate_probability()` as private, there is no reason it shouldn't be public as `estimate_probability()`!

This is minor though, and can be addressed in the next PR that modifies `QubitDevice`.

File: pennylane/plugins/default_qubit.py (hunk 5 above)
Comment (josh146) [reply pennylane/plugins/default_qubit.py:268]: This is now the case I believe!

File: pennylane/plugins/default_qubit.py (hunk 12 above)
Comment (josh146) [reply pennylane/plugins/default_qubit.py:274]: Agree, this should be kept when moved to `QubitDevice`.

File: tests/test_default_qubit.py [hunk 14]
```diff
@@ -1612,3 +1612,61 @@ def test_hermitian(self, theta, phi, varphi, monkeypatch, tol):
...
+            qml.RY(x[1], wires=0)
+            qml.CNOT(wires=[0, 1])
+            return qml.probs(wires=[0, 1])
+
+        prob = qml.QNode(circuit, dev)
+        prob_analytic = qml.QNode(circuit, dev_analytic)
+
+        assert np.isclose(prob(x).sum(), 1, atol=tol, rtol=0)
+        assert np.allclose(prob_analytic(x), prob(x), atol=0.1, rtol=0)
+        assert not np.array_equal(prob_analytic(x), prob(x))
```
Comment (josh146) [inline tests/test_default_qubit.py:1640]: Nice and clear integration test!

File: pennylane/plugins/default_qubit.py (hunk 12 above)
Comment (thisac) [reply pennylane/plugins/default_qubit.py:274]: Sure, I just added the `hasattr` here so that it wouldn't be forgotten when transferring it to `QubitDevice`, but it's a bit useless, and perhaps unnecessary, in `DefaultQubit`.

File: tests/test_default_qubit.py [hunk 15]
```diff
@@ -1612,3 +1612,61 @@ def test_hermitian(self, theta, phi, varphi, monkeypatch, tol):
...
+        dev = qml.device("default.qubit", wires=2, analytic=analytic)
+        monkeypatch.setattr(dev, "analytic_probability", self.mock_analytic_counter)
+
+        # generate samples through `generate_samples` (using 'analytic_probability')
+        dev.generate_samples()
+
+        # should call `analytic_probability` once through `generate_samples`
+        assert self.analytic_counter == 1
+
+    # NOTE: remove test when enforcing `analytic_probability` method for all analytic devices
```
Comment (thisac) [inline tests/test_default_qubit.py:1656]:
```suggestion
    # TODO: remove test when enforcing `analytic_probability` method for all analytic devices
```

File: pennylane/plugins/default_qubit.py (hunk 13 above)
Comment (thisac) [reply pennylane/plugins/default_qubit.py]: I changed it here! Makes sense to have it public (as the other methods). :slightly_smiling_face:
